text/0000-float-semantics.md

```diff
@@ -0,0 +1,282 @@
+
+### Alternative: operations can never produce a signaling NaN
+
+An operation that returns a signaling NaN violates the basic IEEE guarantee that floating-point operations never produce signaling NaNs. However, LLVM considers it legal to fold `x * 1.0` to `x`, so if `x` is a signaling NaN then the multiplication can have that signaling NaN as the output.
+Similarly, `(x: f32) as f64 as f32` may get folded into `x`, so even casts can produce signaling NaNs.
+Given that signaling NaNs are basically useless in Rust since we do not consider the exception status flag to be observable (and hence optimizations can change whether an exception is triggered ot not), there is no point in constraining optimizations just to achieve a guarantee about signaling NaNs.
+
+### Alternative: don't make any guarantee about the signaling/quiet bit ever
+
+This alternative would simplify the spec and make old MIPS hardware compliant.
+The signaling/quiet distinction also basically does not matter since floating-point exception flags are not exposed.
+However, there is one operation in C that can sometimes produce a non-NaN result on *quiet* NaN inputs specifically: [`pow`](https://en.cppreference.com/w/c/numeric/math/pow). For instance, "`pow(+1, exponent)` returns `1` for any `exponent`, even when `exponent` is `NaN`".
+C also says "This specification does not define the behavior of signaling NaNs", and in practice, `pow(1, sNaN)` returns a NaN.
+In other words, `pow` *does* [make a difference](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5a0855c2cb6630256e4407624650c4c9) between quiet and signaling NaN.
+So, if we say that any operation can arbitrairly produce signaling NaNs, then it becomes impossible to rely on `pow`'s property that "`pow(+1, exponent)` returns `1` for any `exponent`".
+Since the quiet/signaling distinction matters, we provide a guarantee which ensures that programs will almost always only ever deal with quiet NaNs (the only safe way to get a signaling NaN is to use `from_bits`).
+
+### Alternative: no strict semantic guarantees
+
+Providing strict IEEE 754-2008 guarantees precludes many transformations, such as turning `a*b + c` into FMA operations (since the result can change due to different rounding).
+We could provide weaker guarantees to allow such transformations.
+(That would also make using the higher-precision x87 instructions conforming with the spec.)
+However, there is no way to bound the effect of such transformations on program behavior: they lead to different outcome observable via `to_bits` and even via pure float operations, which can balloon into arbitrary changes in behavior.
+There is also no "monotonicity of precision": while e.g. an FMA instead of mul-then-add will lead to a higher-precision result for this particular operation, this can in principle lead to a lower-precision result in later computations.
```

Comment:
The classic example here is evaluating `x*x - x*x` as `x.mul_add(x, -x*x)` producing nonzero. This is a bit awkward because `fma` is slow on x86_64 generic so you don't want to use it unless you need it, but fast when target features are enabled (on host or device). One could exclude [`mul_add`](https://doc.rust-lang.org/std/primitive.f64.html#method.mul_add) from `const fn`, but outside of specific cases where symmetry in such expressions are necessary, scientific computing users will usually prefer `mul_add` semantics (both for accuracy and performance). This is especially relevant to GPU targets where FMA is heavily used. It would be a shame to eternally forbid using `mul_add` in `const fn` (thereby degrading its accuracy everywhere and reducing performance when called at run-time). I'm not sure what the right semantic is here, but it's an important niche, I'm concerned that the present semantic may be too strict, and I'm not sure a "fast float type" covers bases in a useful way. (Among other things, trait methods cannot be `const`.)

Reply:
I don't understand which point you are trying to make here. Nobody is suggesting to eternally forbid `mul_add` in `const fn`. This paragraph you are replaying to is about optimizations that automatically introducing FMA into code that didn't use FMA -- the RFC says that that is not permitted.

Reply:
Thanks for clarifying. Lack of support for [`AllowContract`](https://llvm.org/doxygen/classllvm_1_1FastMathFlags.html) is quite painful for people working in scientific computing, and is responsible for a sizeable fraction of the case studies "I ported {physics kernel} from C to Rust and it's much slower" (the rest having easier resolutions, like checked indexing).

The only way to get FMAs with Rust (apart from direct use of `mul_add`, which has bad performance consequences on some architectures and is awkward to program with) is nightly-only (`core_intrinsics` to access the likes of `fadd_fast` and `fmul_fast`, which will always be unsafe sledgehammers). There have been suggestions that new intrinsics like `fadd_with_flags(a, b, flags)` could possibly be stabilized so long as flags like `NoInfs` were unsafe. As I understand it, the proposed semantic here would prevent that from ever being possible (even `AllowContract`, which is the default behavior of many C compilers). Moreover, the strategy of [fast-floats](https://crates.io/crates/fast-floats) would effectively require that all perf-sensitive computation become unsafe (it's hard enough to prove that values are always remain finite, but observable differences between FMA and non-FMA optimization paths are everywhere). This would be a shame since safe Rust can provide immense value to that community.

A different way to ask: is there a way we could leave the door open for a future proposal to bring `AllowContract` semantics in safe Rust?

Reply:
This RFC does not close the door to fast-math operations (Cc https://github.com/rust-lang/rust/issues/21690). It just says they must be opt-in. Otherwise it's entirely orthogonal.

Reply:
> The only way to get FMAs with Rust (apart from direct use of `mul_add`, which has bad performance consequences on some architectures

I've wanted Rust to have a function that translates to the LLVM [`llvm.fmuladd.*` intrinsic](), which is always either a FMA or separate mul and add, no other options.

Reply:
> I see no reason why they must be unsafe

Certainly `fadd_fast`/`fmul_fast` must remain unsafe because `NoNaNs` and `NoInfs` [produce poison](https://llvm.org/docs/LangRef.html#fast-math-flags). The other flags enable target- and optimization-dependent results, but this discussion clarifies that we agree that is okay in safe code.

I was put off by this RFC reading as though NaN-nonuniqueness was a limited carve-out of the convention that runtime and const contexts produce the same results. It looks like [this suggested RFC](https://github.com/rust-lang/rust/issues/77745#issuecomment-1207286013) was never written, but would clarify what are widely varying opinions in the community. I think default semantics leaning toward deterministic is good so long as we don't inhibit an easy/safe/stable way to opt in.

Thanks @programmerjake, I think you meant to link [this](https://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic). The question (for a different thread) is how to expose it.

Reply:
> Certainly fadd_fast/fmul_fast must remain unsafe because NoNaNs and NoInfs [produce poison](https://llvm.org/docs/LangRef.html#fast-math-flags).

Oh sure. But the basic idea of "rustc may optimize assuming the laws of the regular rational numbers" doesn't need unsafe I think. It might just be that the way LLVM exposes it needs unsafe.

> I was put off by this RFC reading as though NaN-nonuniqueness was a limited carve-out of the convention that runtime and const contexts produce the same results. 

We do plan to allow const fn to arbitrarily differ from runtime behavior, but that's a different RFC: https://github.com/rust-lang/rfcs/pull/3352.

This RFC just says that const fn can be non-deterministic when executed at runtime, but the compile-time behavior is still in the set of possible rutnime behaviors. (But "possible" must be interpreted in a spec sense here: the spec allows that behavior to occur. The behavior might never actually happen with current rustc.)